src/megaclient.cpp

```diff
@@ -34,6 +35,80 @@ namespace mega {
 // FIXME: instead of copying nodes, move if the source is in the rubbish to reduce node creation load on the servers
 // FIXME: prevent synced folder from being moved into another synced folder
 
+
+void MegaClientAsyncQueue::push(std::function<void(SymmCipher&)> f)
+{
+    std::lock_guard<std::mutex> g(mMutex);
+    if (mAsyncThreads.empty())
+    {
+        if (f)
+        {
+            f(mZeroThreadsCipher);
+        }
+    }
+    else
+    {
+        mQueue.emplace_back(std::move(f));
+        mConditionVariable.notify_one();
+    }
+}
+
+MegaClientAsyncQueue::MegaClientAsyncQueue(Waiter& w, unsigned threadCount)
+    : mWaiter(w)
+{
+    for (int i = threadCount; i--; )
+    {
+        try
+        {
+            mAsyncThreads.emplace_back([this]()
+            {
+                asyncThreadLoop();
+            });
+        }
+        catch (std::system_error& e)
+        {
+            LOG_err << "Failed to start worker thread: " << e.what();
+            break;
+        }
+    }
+    LOG_debug << "MegaClient Worker threads running: " << mAsyncThreads.size();
+}
+
+MegaClientAsyncQueue::~MegaClientAsyncQueue()
+{
+    clearQueue();
+    push(nullptr);
```

Comment:
Feels quite hacky to me to push a null function to terminate. I'd use a separate boolean for this.

Reply:
I did try that, but it bulked the function up a bit, and didn't look as nice.